src/Perpetuum/Services/EventServices/EventProcessors/OreNPCSpawner.cs

```diff
@@ -0,0 +1,90 @@
+﻿using Perpetuum.ExportedTypes;
+using Perpetuum.Services.EventServices.EventMessages;
+using Perpetuum.Zones;
+using Perpetuum.Zones.NpcSystem.Presences;
+using Perpetuum.Zones.Terrains.Materials.Minerals;
+using System;
+using System.Collections.Generic;
+
+namespace Perpetuum.Services.EventServices.EventProcessors
+{
+    /// <summary>
+    /// </summary>
+    public class OreNPCSpawner : EventProcessor<EventMessage>
+    {
+        private IZone _zone;
+        private readonly IDictionary<MineralNode, DynamicPresence> orePresences = new Dictionary<MineralNode, DynamicPresence>();
+        private ulong _lastTotal;
+
+
+        public OreNPCSpawner(IZone zone)
+        {
+            _zone = zone;
+        }
+
+        private KeyValuePair<MineralNode, DynamicPresence> FindEntryByValue(Presence presence)
+        {
+            foreach (var pair in orePresences)
+            {
+                if (pair.Value == presence)
+                {
+                    return pair;
+                }
+            }
+            return new KeyValuePair<MineralNode, DynamicPresence>(null, null);
```

Comment:
Personally I would use a dedicated response object rather than a null filled KVP. Returning an actual object though invalid can lead to confusion further down the line.

And example of a response class would be;
```
public class NodeEntryResponse {
  public NodeEntryResponse(bool found, MineralNode? node = null, DynamicPresence? presence = null)
  {
    ValidNodeFound = found;
    Node = node;
    Presence = presence;
  }
  public MineralNode? Node {get;}
  public DynamicPresence? Presence {get;}
  public bool ValidNodeFound {get;}

  public KeyValuePair<MineralNode, DynamicPresence> GetAsPair()
  {
    if (!ValidNodeFound) throw new InvalidOperationException("Valid node required!");

    return new KeyValuePair<MineralNode, DynamicPresence>(Node, Presence);
  }
}
```

You can then check the node validity on the class itself and if it's valid, extract the KVP.

I've used ? here to indicate optionality. If they're reference types (and you're not using C# 8) then you don't need to add the question mark. If you do add the question mark, the `return new KeyValuePair...` line must read `Node.Value` and `Presence.Value`